pulpcore/pulpcore/app/tasks/repository.py

```diff
@@ -109,3 +109,40 @@ def delete_version(pk):
 
         # With no more relationships remaining, delete the version.
         version.delete()
+
+
+@shared_task(base=UserFacingTask)
+def create_version(repository_pk, add_content_units, remove_content_units):
+    """
+    Create a new repository version by adding and then removing content units.
+
+    Args:
+        repository_pk (UUID): The primary key for a Repository for which a new Repository Version
+            should be created.
+        add_content_units (list): List of PKs for :class:`~pulpcore.app.models.Content` that
+            should be added to the previous Repository Version for this Repository.
+        remove_content_units (list): List of PKs for:class:`~pulpcore.app.models.Content` that
+            should be removed from the previous Repository Version for this Repository.
+    """
+
+    with transaction.atomic():
+        repository = models.Repository.objects.get(pk=repository_pk)
+        new_version = models.RepositoryVersion(repository=repository)
+        new_version.save()
+        created = models.CreatedResource(content_object=new_version)
+        created.save()
+        repository.last_version = new_version.number
+        repository.save()
+
+    try:
+        for pk in add_content_units:
+            content = models.Content.objects.get(pk=pk)
+            new_version.add_content(content)
+        for pk in remove_content_units:
+            content = models.Content.objects.get(pk=pk)
+            new_version.remove_content(content)
+    except Exception:
+        with transaction.atomic():
+            new_version.delete()
```

Comment:
This cascade deletes all of the RepositoryContent associations too right?

Reply:
I don't think it does. I'll file an issue for this.

Reply:
I think this issue already kind of exists here - https://pulp.plan.io/issues/3226

Reply:
This would be for a non-crash scenario. That issue does not specify that the RepositoryContent table needs to be cleaned up also.

Reply:
Right, I was imagining though that the task would create some sort of cleanup code we could call into. Sorry for the confusion.